docs/user/manual.adoc

```diff
@@ -288,11 +285,37 @@ interface Crate {
     edition: "2015" | "2018";
     /// Dependencies
     deps: Dep[];
+    /// Should this crate be treated as a member of current "workspace".
+    ///
+    /// By default, inferred from the `root_module` (members are the crates which reside
+    /// inside the directory opened in the editor).
+    ///
+    /// Set this too `false` for things like standard library and 3rd party crates to
+    /// enable performance optimizations (rust-analyzer assumes that non-member crates
+    /// don't change).
+    is_workspace_member?: boolean;
+    /// Optionally specify the (super)set of `.rs` files comprising this crate.
+    ///
+    /// By default, rust-analyzer assumes that only files under `root_module.parent` can belong to a crate.
+    /// `include_dirs` are included recursively, unless a subdirectory is in `exclude_dirs`.
+    ///
+    /// Different crates can share the same `source`.
+
+    /// If two crates share an `.rs` file in common, they *must* have the same `source`.
+    /// rust-analyzer assumes that files from one source can't refer to files in another source.
+    source?: {
+        include_dirs: string[],
+        exclude_dirs: string[],
+    },
```

Comment:
This is the thing to replace `roots`. We still have a constraint that we want to know the set of `.rs` file up-front (otherwise, to server, eg, goto definition for `foo.rs`, we would have to ask every carte "do you include `foo.rs` by any chance?").

It's unclear how to best represent the "set of files on the file system", I went for the simplest non-trivial solution for the time being (went = store include/exclude lists from Dart analysis server). It shouldn't be a problem to support a literal listing of all files or globbing, if we need that.

Reply:
What happens if two crates contain the same `.rs` file, but don't have the same `source`?